src/server/pkg/hashtree/error.go

```diff
@@ -0,0 +1,37 @@
+package hashtree
+
+import "fmt"
+
+// hashTreeError is a custom error type, returned by HashTree's methods
+type hashTreeError struct {
+	code ErrCode
+	s    string
+}
+
+func (e *hashTreeError) Error() string {
+	return e.s
+}
+
+// Code returns the "error code"  of 'err' if it was returned by one of the
+// HashTree methods, or "Unknown" if 'err' was emitted by some other function
+// (error codes are defined in interface.go)
+func Code(err error) ErrCode {
```

Comment:
Interesting approach to error handling.  I think the pattern I've most commonly seen goes like:

```
if err != nil {
  if err, ok := err.(hashTreeError); ok {
    switch err.Code {
      // deal with different error codes...
    }
  } else {
    // it's not a hashTreeError
  }
}
```